File: recommendation_system/cf_example.R
library(reshape)
setwd("/Users/jacquessham/Github/machine_learning_notes/Data")


data.germany <- read.csv("lastfm-matrix-germany.csv")

### Item Based Collaborative Filtering
## Take out user column
data.germany.ibs <- (data.germany[,!(names(data.germany) %in% c("user"))])

## Cosine similarity function
getCosine <- function(x,y) 
{
  this.cosine <- sum(x*y) / (sqrt(sum(x*x)) * sqrt(sum(y*y)))
  return(this.cosine)
}

## Create a p x p similarity matrix
## Create a placeholder dataframe listing item vs. item
data.germany.ibs.similarity  <- matrix(NA, nrow=ncol(data.germany.ibs),ncol=ncol(data.germany.ibs),dimnames=list(colnames(data.germany.ibs),colnames(data.germany.ibs)))

## Find the similarities column by column
for(i in 1:ncol(data.germany.ibs)) {
  # Loop through the columns for each column
  for(j in 1:ncol(data.germany.ibs)) {
    # Fill in placeholder with cosine similarities
    data.germany.ibs.similarity[i,j] <- getCosine(as.matrix(data.germany.ibs[i]),as.matrix(data.germany.ibs[j]))
  }
}

## Convert back to dataframe
data.germany.ibs.similarity <- as.data.frame(data.germany.ibs.similarity)

## Declare a new matrix on top 10 similarity against each option
## Get the top 10 neighbours for each
data.germany.neighbours <- matrix(NA, nrow=ncol(data.germany.ibs.similarity),ncol=11,dimnames=list(colnames(data.germany.ibs.similarity)))

## Convert number to name
for(i in 1:ncol(data.germany.ibs)) {
  data.germany.neighbours[i,] <- (t(head(n=11,rownames(data.germany.ibs.similarity[order(data.germany.ibs.similarity[,i],decreasing=TRUE),][i]))))
}

##This means for those listening to Abba we would recommend Madonna and Robbie Williams.
##Likewise for people listening to ACDC we would recommend the Red Hot Chilli Peppers and Metallica.

### User-based recommendation system
# Process:
# Choose a product, see if the user purchased a product
# Get the similarities of that product's top 10 neighbours
# Get the purchase record of that user of the top 10 neighbours
# Do the formula: sumproduct(purchaseHistory, similarities)/sum(similarities)

# Lets make a helper function to calculate the scores
getScore <- function(history, similarities)
{
  x <- sum(history*similarities)/sum(similarities)
  x
}

# A placeholder matrix
holder <- matrix(NA, nrow=nrow(data.germany),ncol=ncol(data.germany)-1,dimnames=list((data.germany$user),colnames(data.germany[-1])))

# Loop through the users (rows)
for(i in 1:nrow(holder)) 
{
  # Loops through the products (columns)
  for(j in 1:ncol(holder)) 
  {
    # Get the user's name and th product's name
    # We do this not to conform with vectors sorted differently 
    user <- rownames(holder)[i]
    product <- colnames(holder)[j]
    
    # We do not want to recommend products you have already consumed
    # If you have already consumed it, we store an empty string
    if(as.integer(data.germany[data.germany$user==user,product]) == 1)
    { 
      holder[i,j]<-""
    } else {
      
      # We first have to get a product's top 10 neighbours sorted by similarity
      topN<-((head(n=11,(data.germany.ibs.similarity[order(data.germany.ibs.similarity[,product],decreasing=TRUE),][product]))))
      topN.names <- as.character(rownames(topN))
      topN.similarities <- as.numeric(topN[,1])
      
      # Drop the first one because it will always be the same song
      topN.similarities<-topN.similarities[-1]
      topN.names<-topN.names[-1]
      
      # We then get the user's purchase history for those 10 items
      topN.purchases<- data.germany[,c("user",topN.names)]
      topN.userPurchases<-topN.purchases[topN.purchases$user==user,]
      topN.userPurchases <- as.numeric(topN.userPurchases[!(names(topN.userPurchases) %in% c("user"))])
      
      # We then calculate the score for that product and that user
      holder[i,j]<-getScore(similarities=topN.similarities,history=topN.userPurchases)
      
    } # close else statement
  } # end product for loop   
} # end user for loop

# Output the results to a file
data.germany.user.scores <- holder
write.csv(file="final-user-scores.csv",data.germany.user.scores)

# Lets make our recommendations pretty
data.germany.user.scores.holder <- matrix(NA, nrow=nrow(data.germany.user.scores),ncol=100,dimnames=list(rownames(data.germany.user.scores)))
for(i in 1:nrow(data.germany.user.scores)) 
{
  data.germany.user.scores.holder[i,] <- names(head(n=100,(data.germany.user.scores[,order(data.germany.user.scores[i,],decreasing=TRUE)])[i,]))
}

# Write output to file
write.csv(file="final-user-recommendations.csv",data.germany.user.scores.holder)

File: recommendation_system/cf_example.py
import numpy as np
import pandas as pd


# Function to 
def cos_similarity(x,y):
    sum_x2 = np.sum(np.square(x))
    sum_y2 = np.sum(np.square(y))
    sum_xy = np.sum([x*y for x,y in zip(x,y)])
    return sum_xy/np.sqrt(sum_x2*sum_y2)

# Read data and get the list of singers available
print('Reading dataset...')
df_lastfm = pd.read_csv('../Data/lastfm-matrix-germany.csv')
singers = df_lastfm.drop('user', axis=1).columns

# Gather parameters
save_matrix = True
save_result = True
num_recommend = 10

# Loop over each singer and find
print('Calculating similarity scores...')
similarity_singers = []
for i in singers:
	temp = []
	for j in singers:
		temp.append(cos_similarity(df_lastfm[i],df_lastfm[j]))
	similarity_singers.append(temp)

# Re-frame it into Pandas DataFrame
print('Preparing the Similarity Matrix...')
similarity_singers = pd.DataFrame(similarity_singers, columns=singers)
# Add column of singers to make it matrix-like to save
similarity_singers['singers'] = singers
# Rearrange columns order
similarity_singers = similarity_singers[['singers']+singers.tolist()]
print('Similarity Matrix Prepared!')
# Save matrix to csv
if save_matrix:
	print('Saving Similarity Matrix...')
	similarity_singers.to_csv('Results/item_cf_similarity_matrix.csv', index=False)
	print('Similarity Matrix sucessfully saved!')


# Convert similarity to recommendation (Number convert to singer names)
def convert_name(df, col_singers, col_score, top_n):
	df = df.sort_values(col_score).head(top_n+1)
	col_result = df[col_singers].iloc[1:].tolist()
	return col_result

print('Converting Similarity Matrix to result...')
result = {}
for col in similarity_singers.columns:
	if col != 'singers':
		result[col] = convert_name(similarity_singers, 'singers',
			                       col, num_recommend)
result = pd.DataFrame(result)
print('Result Prepared!')
if save_result:
	print('Saving result...')
	result.to_csv('Results/it_cf_result.csv', index=False)
	print('Result successfully saved!')

